### PR TITLE
feat(sbom): normalize SBOMs for deterministic builds

### PR DESCRIPTION
**Pull request by leodido**
## Overview

Makes SBOM generation deterministic by normalizing timestamps and UUIDs after Syft generates them. This enables reproducible builds where the same source produces the same artifact SHA256.

Part of https://linear.app/ona-team/issue/CLC-2096/prevent-attestation-overwrites-in-concurrent-builds
Part of https://linear.app/ona-team/issue/CLC-2097/improve-builds-determinism

### Implementation (`pkg/leeway/sbom.go`)

1. **`getGitCommitTimestamp()`**: Retrieves deterministic timestamp from git commit
   - Respects `SOURCE_DATE_EPOCH` environment variable for reproducible builds
   - Falls back to git commit timestamp
   - Logs warning if `SOURCE_DATE_EPOCH` is invalid

2. **`generateDeterministicUUID()`**: Generates UUIDv5 from content
   - Uses RFC 4122 DNS namespace UUID
   - SHA-1 based for determinism

3. **`normalizeCycloneDX()`**: Normalizes CycloneDX SBOM
   - Replaces timestamp with git commit time
   - Generates deterministic UUID from content

4. **`normalizeSPDX()`**: Normalizes SPDX SBOM
   - Replaces timestamp with git commit time
   - Generates deterministic UUID in documentNamespace

5. **Updated `writeSBOM()`**: Calls normalizers after SBOM generation
   - Fails fast if deterministic timestamp cannot be obtained
   - Provides clear error message with troubleshooting hints

### Tests (`pkg/leeway/sbom_normalize_test.go`)

Comprehensive test coverage including:
- UUID generation (determinism, format validation)
- Git commit timestamp (SOURCE_DATE_EPOCH, git fallback)
- CycloneDX normalization (happy path, malformed input, file errors)
- SPDX normalization (happy path, malformed input, file errors)

### Problem
SBOMs contained non-deterministic fields:
- Current timestamp (changes on every build)
- Random UUIDs (different on every build)

This prevented:
- Reproducible builds (same source → different artifact hash)
- Reliable caching
- Attestation verification

### Solution
Post-process SBOMs after Syft generates them to replace non-deterministic values with deterministic ones based on git commit metadata.

### Why Not Wait for Upstream?
- Syft has an [open PR](https://github.com/anchore/syft/pull/3932) for this feature, but it's not merged yet
- Our post-processing approach works with any Syft version
- Both approaches are compatible (we can remove normalization once Syft supports it natively)

## Format-Specific Behavior

- **CycloneDX**: Normalized (has timestamps and UUIDs)
- **SPDX**: Normalized (has timestamps and UUIDs)
- **Syft JSON**: No normalization needed (naturally deterministic - no timestamp field, no random UUIDs)

See [anchore/syft#3931](https://github.com/anchore/syft/issues/3931) for upstream discussion.

### Unit Tests
```bash
go test -v ./pkg/leeway -run "TestGenerateDeterministicUUID|TestGetGitCommitTimestamp|TestNormalizeCycloneDX|TestNormalizeSPDX"
```

All tests pass with coverage for normal operation and error conditions.

### Manual Verification
```bash
# Build twice from scratch
rm -rf /tmp/leeway-cache
leeway build //:helloworld -Dversion=test1 --save /tmp/build1.tar.gz

rm -rf /tmp/leeway-cache
leeway build //:helloworld -Dversion=test1 --save /tmp/build2.tar.gz

# Verify SBOMs are identical
tar -xzf /tmp/build1.tar.gz ./sbom.cdx.json && mv sbom.cdx.json sbom1.cdx.json
tar -xzf /tmp/build2.tar.gz ./sbom.cdx.json && mv sbom.cdx.json sbom2.cdx.json
diff sbom1.cdx.json sbom2.cdx.json  # No differences

# Verify timestamp uses git commit time
cat sbom1.cdx.json | jq '.metadata.timestamp'
# Output: "2025-11-17T09:11:11Z" (matches git commit timestamp)

# Verify UUIDs are deterministic
cat sbom1.cdx.json | jq '.serialNumber'
cat sbom2.cdx.json | jq '.serialNumber'
# Both output: "urn:uuid:7cfdf9a9-1f08-55b0-9fbb-23cee00912fe"
```

## Error Handling

Fails fast with clear error message if deterministic timestamp cannot be obtained:
```
failed to get deterministic timestamp for SBOM normalization (commit: abc123): failed to get commit timestamp: ...
Ensure git is available and the repository is not a shallow clone, or set SOURCE_DATE_EPOCH environment variable
```

## Breaking Changes

None. This is a non-breaking enhancement that makes builds more deterministic.

Co-authored-by: Ona <no-reply@ona.com>

### CODE REVIEW

**Review by geropl — APPROVED** [review]
Changes and tests LGTM! :heavy_check_mark:

File: pkg/leeway/sbom.go [hunk 1]
```diff
@@ -96,6 +102,142 @@ func GetSBOMParallelism(sbomConfig WorkspaceSBOM) int {
...
+	if err != nil {
+		return fmt.Errorf("failed to marshal SBOM for hashing: %w", err)
+	}
+	contentHash := sha256.Sum256(normalizedForHash)
+	deterministicUUID := generateDeterministicUUID(contentHash[:])
+
+	// Replace UUID in documentNamespace
+	if originalNamespace != "" {
+		// Extract base URL and replace UUID at the end
+		lastDash := strings.LastIndex(originalNamespace, "-")
```
Comment (corneliusludmann) [inline pkg/leeway/sbom.go]: Do you think it would make sense to validate the format of `originalNamespace` here?

- Assumes UUID is at the end of the namespace
- Doesn't validate that what it's replacing is actually a UUID
- Could fail silently if the namespace format is unexpected
- No test coverage for edge cases (namespace without UUID, UUID in middle, etc.)

Something like this:

```
uuidPattern := regexp.MustCompile(`[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}`)
if uuidPattern.MatchString(originalNamespace) {
    originalNamespace = uuidPattern.ReplaceAllString(originalNamespace, deterministicUUID)
}
```

**Review by corneliusludmann — APPROVED** [review]
Added 2 comments with improvements (nothing major). Approve to unblock.

File: pkg/leeway/sbom.go [hunk 2]
```diff
@@ -96,6 +102,142 @@ func GetSBOMParallelism(sbomConfig WorkspaceSBOM) int {
...
+		timestamp, err := strconv.ParseInt(epoch, 10, 64)
+		if err == nil {
+			return time.Unix(timestamp, 0).UTC(), nil
+		}
+		// Log warning but continue to git fallback
+		log.WithError(err).WithField("SOURCE_DATE_EPOCH", epoch).Warn("Invalid SOURCE_DATE_EPOCH, falling back to git commit timestamp")
+	}
+
+	// Get commit timestamp from git
+	cmd := exec.Command("git", "show", "-s", "--format=%ct", commit)
```
[the remark this answers is not in the document]
Comment (leodido) [reply pkg/leeway/sbom.go]: Done
